packages/compiler-cli/src/ngtsc/annotations/directive/src/query_functions.ts

```diff
@@ -0,0 +1,110 @@
+ * first one. The query is then analyzed and its resolved metadata is returned.
+ *
+ * @returns Resolved query metadata, or null if no query is declared.
+ */
+export function tryParseSignalQueryFromInitializer(
+    member: Pick<ClassMember, 'name'|'value'>, reflector: ReflectionHost,
+    isCore: boolean): {name: QueryFunctionName, metadata: R3QueryMetadata}|null {
+  const query = tryParseInitializerApiMember(queryFunctionNames, member, reflector, isCore);
+  if (query === null) {
+    return null;
+  }
+
+  const isSingleQuery = query.apiName === 'viewChild' || query.apiName === 'contentChild';
+  const predicateNode = <ts.Expression|undefined>query.call.arguments[0];
+  if (predicateNode === undefined) {
+    throw new FatalDiagnosticError(
+        ErrorCode.VALUE_HAS_WRONG_TYPE, query.call, 'No locator specified.');
+  }
+
+  const optionsNode = <ts.Expression|undefined>query.call.arguments[1];
+  if (optionsNode !== undefined && !ts.isObjectLiteralExpression(optionsNode)) {
+    throw new FatalDiagnosticError(
+        ErrorCode.VALUE_HAS_WRONG_TYPE, optionsNode, 'Argument needs to be an object literal.');
+  }
+  const options = optionsNode && reflectObjectLiteral(optionsNode);
+  const read = options?.has('read') ? parseReadOption(options.get('read')!) : null;
+  const descendants = options?.has('descendants') ?
+      parseDescendantsOption(options.get('descendants')!) :
+      defaultDescendantsValue(query.apiName);
+
+  return {
+    name: query.apiName,
+    metadata: {
+      isSignal: true,
+      propertyName: member.name,
+      static: false,
+      emitDistinctChangesOnly: true,
+      predicate: parseLocator(predicateNode, reflector),
+      first: isSingleQuery,
+      read,
+      descendants,
+    },
+  };
+}
+
+/** Parses the locator/predicate of the query. */
+function parseLocator(expression: ts.Expression, reflector: ReflectionHost): string[]|
+    MaybeForwardRefExpression<o.Expression> {
+  // Attempt to unwrap `forwardRef` calls.
+  const unwrappedExpression = tryUnwrapForwardRef(expression, reflector);
+  if (unwrappedExpression !== null) {
+    expression = unwrappedExpression;
+  }
+
+  if (ts.isStringLiteralLike(expression)) {
+    return [expression.text];
+  }
+
+  return createMayBeForwardRefExpression(
+      new o.WrappedNodeExpr(expression),
+      unwrappedExpression !== null ? ForwardRefHandling.Unwrapped : ForwardRefHandling.None);
+}
+
+/** Parses the `read` option of a query. */
+function parseReadOption(value: ts.Expression): o.Expression {
+  return new o.WrappedNodeExpr(value);
+}
+
+/** Parses the `descendants` option of a query. */
+function parseDescendantsOption(value: ts.Expression): boolean {
+  if (value.kind === ts.SyntaxKind.TrueKeyword) {
+    return true;
+  } else if (value.kind === ts.SyntaxKind.FalseKeyword) {
+    return false;
+  }
+  throw new FatalDiagnosticError(
+      ErrorCode.VALUE_HAS_WRONG_TYPE, value, `Expected "descendants" option to be a boolean.`);
```

Comment:
nit: to avoid confusion that this is not just a boolean type, but a boolean literal specifically:

```suggestion
      ErrorCode.VALUE_HAS_WRONG_TYPE, value, `Expected "descendants" option to be a boolean literal.`);
```